File: ipad-app/Lasso.swift
import CoreGraphics

/// 올가미 폴리곤 판정.
enum Lasso {

    /// 점이 폴리곤 안에 있는지 (even-odd ray casting).
    static func contains(_ polygon: [CGPoint], _ p: CGPoint) -> Bool {
        var inside = false
        var j = polygon.count - 1
        for i in 0..<polygon.count {
            let a = polygon[i], b = polygon[j]
            if (a.y > p.y) != (b.y > p.y),
               p.x < (b.x - a.x) * (p.y - a.y) / (b.y - a.y) + a.x {
                inside.toggle()
            }
            j = i
        }
        return inside
    }

    /// 폴리곤에 점이 하나라도 들어간 획들의 인덱스.
    static func strokesInside(_ strokes: [Stroke], polygon: [CGPoint]) -> Set<Int> {
        guard polygon.count >= 3 else { return [] }
        var result: Set<Int> = []
        for (i, stroke) in strokes.enumerated() {
            if stroke.points.contains(where: { contains(polygon, $0.location) }) {
                result.insert(i)
            }
        }
        return result
    }
}

File: flutter-app/lib/canvas/lasso.dart
import 'dart:ui';

import 'stroke.dart';

/// 올가미 폴리곤 판정.

/// 점이 폴리곤 안에 있는지 (even-odd ray casting).
bool pointInPolygon(Offset p, List<Offset> poly) {
  var inside = false;
  for (var i = 0, j = poly.length - 1; i < poly.length; j = i++) {
    final a = poly[i], b = poly[j];
    if ((a.dy > p.dy) != (b.dy > p.dy)) {
      final x = a.dx + (p.dy - a.dy) / (b.dy - a.dy) * (b.dx - a.dx);
      if (p.dx < x) inside = !inside;
    }
  }
  return inside;
}

/// 폴리곤에 점이 하나라도 들어간 획들의 인덱스.
Set<int> strokesInside(List<Stroke> strokes, List<Offset> lasso) {
  if (lasso.length < 3) return {};
  return {
    for (var i = 0; i < strokes.length; i++)
      if (strokes[i].points.any((p) => pointInPolygon(p.offset, lasso))) i,
  };
}
